### PR TITLE
WIP: Fix name collisions

### PR DESCRIPTION
**Pull request by williammartin**
Opening for discussion regarding: https://github.com/storyscript/cli/issues/283

There are many ways we could tackle this problem including:
 - The one presented
 - Adding a similar query on server side to get app by name
 - Actually make names unique

And maybe others.

This PR (incomplete, just for discussion) seems like the easiest way to resolve _already_ broken apps, and has the added benefit of additionally resolving other apps that are in a broken state because the name was invalid (e.g. uppercase characters, or hyphens).

This frustrates me at least once a week when I choose a name that:
 - has been used before
 - isn't acceptable for DNS

### CODE REVIEW

**Comment by judepereira** [comment]
@williammartin I like the approach. Can we add one more thing? Along with getting apps by their names (which you've added), can you also retrieve their correct DNS names so that we can print the right ones post a deployment in the CLI?

**Comment by williammartin** [comment]
> The REAL problem is that we never save the actual app UUID for a user

By save I'm guessing you mean in the `story.yml`?

**Comment by wilzbach** [comment]
> By save I'm guessing you mean in the story.yml?

Not necessarily, it could also be in a user's app data. I think in the repository is better because the configs will survive a migration to a different harddrive (and can be checked into git).

We will have to add a `story.lock` file soon (see e.g. https://github.com/storyscript/hub-sdk-python/issues/27), so this another good candidate.

**Comment by williammartin** [comment]
I'm persuaded that your suggestion is better long term if we do it right. I don't really like the idea of having the uuid in `story.yml` but it's part of a bigger thing I think I want to untangle (reuse of app names). I suppose a solution I could see would be something akin to `terraform.json` and `terraform.state` files (one is declarative configuration, one is the results of applying that configuration).

Assuming we want to get people that are currently broken out of this state, and we think it's worth the effort it still feels like we may also need similar logic to this PR as an intermediary step. (i.e. try to find app uuid, only find name, fetch app uuid _by name_, store it for later). Do you have any other thoughts on this?

I don't want to boil the ocean on this - just quickly resolve this pain that I'm feeling regularly.

Cheers

**Comment by wilzbach** [comment]
> Assuming we want to get people that are currently broken out of this state, and we think it's worth the effort it still feels like we may also need similar logic to this PR as an intermediary step. (i.e. try to find app uuid, only find name, fetch app uuid by name, store it for later). Do you have any other thoughts on this?

Well, it does boil down to making sure that all app names are _unique_. Rename those who conflict. IMHO there are few enough apps out there that this can be done directly without a migration. If you want to do this with a migration period, two solutions:
- store UUID <-> app-name in a global app cache before the rename (global has the advantage that it's not binding us, e.g. it's the easiest to remove/change)
- let CLI detect app-names with conflicts and trigger a warning (this won't be very effective though)

tl;dr: either make app names unique (like e.g. Heroku) or start storing the UUID (like e.g. AWS). You really need it.

I just think that if you want to go down the UUID route, you need to do it anyhow, so why not directly here?
I don't feel strongly about it which is why I withdrew my review.

**Comment by williammartin** [comment]
I think we're talking about the same thing here now.

I am in favour of UUID to remove the constraint on app names being globally unqiue.
I believe we still need the logic to convert app name -> UUID (rather than dns -> UUID) for anyone that is currently broken.

Thanks!
